src/Razor/test/Microsoft.AspNetCore.Razor.LanguageServer.Test/WorkspaceDiagnosticRefreshTest.cs

```diff
@@ -0,0 +1,188 @@
+            });
+
+        await testAccessor.WaitForRefreshAsync();
+
+        await projectSnapshotManager.UpdateAsync(
+            static updater =>
+            {
+                var project = (ProjectSnapshot)updater.GetProjects().First();
+                var directory = Path.GetDirectoryName(project.FilePath);
+                Assert.NotNull(directory);
+
+                var filePath = Path.Combine(directory, "document.razor");
+                updater.CreateAndAddDocument(project, filePath);
+            });
+
+        await testAccessor.WaitForRefreshAsync();
+
+        clientConnection.Verify(
+            c => c.SendNotificationAsync(Methods.WorkspaceDiagnosticRefreshName, It.IsAny<CancellationToken>()),
+            Times.Exactly(2));
+    }
+
+    [Fact]
+    public async Task WorkspaceRefreshNotSent_ClientDoesNotSupport()
+    {
+        var projectSnapshotManager = CreateProjectSnapshotManager();
+        var clientConnection = new StrictMock<IClientConnection>();
+
+        using var publisher = new WorkspaceDiagnosticsRefresher(
+            projectSnapshotManager,
+            new TestClientCapabilitiesService(new()
+            {
+                Workspace = new()
+                {
+                    Diagnostics = new()
+                    {
+                        RefreshSupport = false
+                    }
+                }
+            }),
+            clientConnection.Object,
+            s_delay);
+
+        var testAccessor = publisher.GetTestAccessor();
+
+        await projectSnapshotManager.UpdateAsync(
+            static updater =>
+            {
+                updater.CreateAndAddProject("C:/path/to/project.csproj");
+            });
+
+        await testAccessor.WaitForRefreshAsync();
+
+        clientConnection
+            .Verify(c => c.SendNotificationAsync(It.IsAny<string>(), It.IsAny<CancellationToken>()),
+                    Times.Never);
+    }
+
+    [Fact]
+    public async Task WorkspaceRefreshNotSent_RefresherDisposed()
+    {
+        var projectSnapshotManager = CreateProjectSnapshotManager();
+        var clientConnection = new StrictMock<IClientConnection>();
+
+        var publisher = new WorkspaceDiagnosticsRefresher(
+            projectSnapshotManager,
+            new TestClientCapabilitiesService(new()
+            {
+                Workspace = new()
+                {
+                    Diagnostics = new()
+                    {
+                        RefreshSupport = false
+                    }
+                }
+            }),
+            clientConnection.Object,
+            s_delay);
+
+        var testAccessor = publisher.GetTestAccessor();
+
+        publisher.Dispose();
+
+        await projectSnapshotManager.UpdateAsync(
+            static updater =>
+            {
+                updater.CreateAndAddProject("C:/path/to/project.csproj");
+            });
+
+        await testAccessor.WaitForRefreshAsync();
```

Comment:
Since `publisher` is disposed above, will this every return?

Reply:
I believe so. The underlying ABWQ method doesn't seem to check for cancellation, it just returns the task. That task is completed when this hits